# 🎬 IMDb Top 1000 Movies — Exploratory Data Analysis

**Goal:** Answer 5 real questions about the best movies ever made using pandas and numpy.

**Dataset:** IMDb Top 1000 Movies (Kaggle)

**Questions we'll answer:**
1. Which genres have the highest average IMDb rating?
2. Who are the top 10 directors by number of movies?
3. Is there a correlation between votes and rating?
4. How have average ratings changed across decades?
5. Which movies have missing revenue — is there a pattern?

## 📦 Imports

In [1]:
import pandas as pd
import numpy as np

print("pandas version:", pd.__version__)
print("numpy version:", np.__version__)

pandas version: 3.0.3
numpy version: 2.4.6


## 📥 Load Data

In [2]:
df = pd.read_csv("imdb_top_1000.csv")
print("Shape:", df.shape)
print("\nColumns:", df.columns.tolist())
print("\nFirst 3 rows:")
df.head(3)

Shape: (1000, 16)

Columns: ['Poster_Link', 'Series_Title', 'Released_Year', 'Certificate', 'Runtime', 'Genre', 'IMDB_Rating', 'Overview', 'Meta_score', 'Director', 'Star1', 'Star2', 'Star3', 'Star4', 'No_of_Votes', 'Gross']

First 3 rows:


,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
0,https://m.media-amazon.com/images/M/MV5BMDFkYT...,The Shawshank Redemption,1994,A,142 min,Drama,9.3,Two imprisoned men bond over a number of years...,80.0,Frank Darabont,Tim Robbins,Morgan Freeman,Bob Gunton,William Sadler,2343110,"28,341,469"
1,https://m.media-amazon.com/images/M/MV5BM2MyNj...,The Godfather,1972,A,175 min,"Crime, Drama",9.2,An organized crime dynasty's aging patriarch t...,100.0,Francis Ford Coppola,Marlon Brando,Al Pacino,James Caan,Diane Keaton,1620367,"134,966,411"
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,"534,858,444"


## 🔍 Data Audit

In [3]:
print("=== NULL COUNTS ===")
print(df.isnull().sum())
print("\n=== DATA TYPES ===")
print(df.dtypes)
print("\n=== BASIC STATS ===")
df[['IMDB_Rating', 'No_of_Votes', 'Meta_score']].describe()

=== NULL COUNTS ===
Poster_Link        0
Series_Title       0
Released_Year      0
Certificate      101
Runtime            0
Genre              0
IMDB_Rating        0
Overview           0
Meta_score       157
Director           0
Star1              0
Star2              0
Star3              0
Star4              0
No_of_Votes        0
Gross            169
dtype: int64

=== DATA TYPES ===
Poster_Link          str
Series_Title         str
Released_Year        str
Certificate          str
Runtime              str
Genre                str
IMDB_Rating      float64
Overview             str
Meta_score       float64
Director             str
Star1                str
Star2                str
Star3                str
Star4                str
No_of_Votes        int64
Gross                str
dtype: object

=== BASIC STATS ===


,IMDB_Rating,No_of_Votes,Meta_score
count,1000.000000,1.000000e+03,843.000000
mean,7.949300,2.736929e+05,77.971530
std,0.275491,3.273727e+05,12.376099
min,7.600000,2.508800e+04,28.000000
25%,7.700000,5.552625e+04,70.000000
50%,7.900000,1.385485e+05,79.000000
75%,8.100000,3.741612e+05,87.000000
max,9.300000,2.343110e+06,100.000000


## 🧹 Data Cleaning

In [4]:
df['Released_Year'] = pd.to_numeric(df['Released_Year'], errors='coerce')
df['Runtime'] = df['Runtime'].str.replace(" min", "").astype(int)
df['Gross'] = df['Gross'].str.replace(",", "").astype(float)
df['Decade'] = (df['Released_Year'] // 10 * 10).astype("Int64")
print("Cleaning Done ✅")
print(df[['Series_Title', 'Runtime', 'Gross', 'Decade']].head())

Cleaning Done ✅
               Series_Title  Runtime        Gross  Decade
0  The Shawshank Redemption      142   28341469.0    1990
1             The Godfather      175  134966411.0    1970
2           The Dark Knight      152  534858444.0    2000
3    The Godfather: Part II      202   57300000.0    1970
4              12 Angry Men       96    4360000.0    1950


---
## ❓ Question 1 — Which genres have the highest average IMDb rating?

In [5]:
genre_df = df['Genre'].str.split(", ").explode()
genre_rating = pd.DataFrame({
    "Genre" : genre_df,
    "IMDB_Rating" : df.loc[genre_df.index, "IMDB_Rating"]
})
genre_avg = genre_rating.groupby("Genre")["IMDB_Rating"].mean().sort_values(ascending=False)
print(genre_avg)
print("\nTop 3 genres:", genre_avg.head(3).index.tolist())

Genre
War          8.013725
Western      8.000000
Film-Noir    7.989474
Sci-Fi       7.977612
Mystery      7.967677
Drama        7.959392
Crime        7.954545
History      7.953571
Adventure    7.952041
Action       7.948677
Musical      7.947059
Biography    7.935780
Fantasy      7.931818
Animation    7.930488
Sport        7.926316
Romance      7.925600
Music        7.914286
Family       7.912500
Thriller     7.909489
Comedy       7.903433
Horror       7.887500
Name: IMDB_Rating, dtype: float64

Top 3 genres: ['War', 'Western', 'Film-Noir']


---
## ❓ Question 2 — Who are the top 10 directors by number of movies?

In [6]:
top_directors = df["Director"].value_counts().head(10)
print("Top 10 Directors by Movie Count:")
print(top_directors)
director_avg_rating = df.groupby("Director")["IMDB_Rating"].mean()
top_director_names = top_directors.index.tolist()
print("\nTheir average IMDb ratings:")
print(director_avg_rating.loc[top_director_names].round(2))

Top 10 Directors by Movie Count:
Director
Alfred Hitchcock     14
Steven Spielberg     13
Hayao Miyazaki       11
Martin Scorsese      10
Akira Kurosawa       10
Stanley Kubrick       9
Billy Wilder          9
Woody Allen           9
Christopher Nolan     8
Quentin Tarantino     8
Name: count, dtype: int64

Their average IMDb ratings:
Director
Alfred Hitchcock     8.01
Steven Spielberg     8.03
Hayao Miyazaki       8.02
Martin Scorsese      8.17
Akira Kurosawa       8.22
Stanley Kubrick      8.23
Billy Wilder         8.14
Woody Allen          7.79
Christopher Nolan    8.46
Quentin Tarantino    8.18
Name: IMDB_Rating, dtype: float64


---
## ❓ Question 3 — Is there a correlation between number of votes and IMDb rating?

In [7]:
correlation = df['No_of_Votes'].corr(df['IMDB_Rating'])
print(f"Correlation between Votes and Rating: {correlation:.4f}")
df["Vote_Range"] = pd.cut(df["No_of_Votes"],
                          bins=[0, 50000, 200000, 500000, 2500000],
                          labels=["Low", "Medium", "High", "Very High"])
vote_rating = df.groupby("Vote_Range", observed=True)["IMDB_Rating"].mean().round(3)
print("\nAverage rating by vote range:")
print(vote_rating)
print("\nMovie count per range:")
print(df["Vote_Range"].value_counts().sort_index())

Correlation between Votes and Rating: 0.4950

Average rating by vote range:
Vote_Range
Low          7.898
Medium       7.896
High         7.903
Very High    8.178
Name: IMDB_Rating, dtype: float64

Movie count per range:
Vote_Range
Low          207
Medium       386
High         224
Very High    183
Name: count, dtype: int64


---
## ❓ Question 4 — How have average ratings changed across decades?

In [8]:
decade_stats = df.groupby("Decade", observed=True).agg(
    average_rating = ("IMDB_Rating", "mean"),
    movie_count = ("Series_Title","count")
).round(2)
print(decade_stats)

        average_rating  movie_count
Decade                             
1920              8.13           11
1930              7.97           24
1940              8.03           35
1950              8.06           56
1960              7.97           73
1970              7.97           76
1980              7.95           89
1990              7.96          150
2000              7.90          237
2010              7.92          242
2020              8.13            6


---
## ❓ Question 5 — Which movies have missing revenue — is there a pattern?

In [9]:
df["Has_Gross"] = df["Gross"].isnull()
print("=== Average rating: movies WITH vs WITHOUT gross data ===")
print(df.groupby("Has_Gross")["IMDB_Rating"].mean().round(3))
print("\n=== Missing gross by decade ===")
missing_by_decade = df[df["Has_Gross"] == True].groupby("Decade", observed=True)["Series_Title"].count()
print(missing_by_decade)
print("\n=== Missing gross % by decade ===")
total_by_decade = df.groupby("Decade")["Series_Title"].count()
missing_pct = (missing_by_decade / total_by_decade * 100).round(1)
print(missing_pct)

=== Average rating: movies WITH vs WITHOUT gross data ===
Has_Gross
False    7.947
True     7.961
Name: IMDB_Rating, dtype: float64

=== Missing gross by decade ===
Decade
1920     3
1930    11
1940    15
1950    25
1960    24
1970    17
1980     9
1990    10
2000    22
2010    27
2020     6
Name: Series_Title, dtype: int64

=== Missing gross % by decade ===
Decade
1920     27.3
1930     45.8
1940     42.9
1950     44.6
1960     32.9
1970     22.4
1980     10.1
1990      6.7
2000      9.3
2010     11.2
2020    100.0
Name: Series_Title, dtype: float64


---
## 📋 Key Findings

1. **Genre** — War, Western and Film-Noir have the highest average ratings. Popular genres like Action and Comedy rank near the bottom.

2. **Directors** — Christopher Nolan has the highest average rating (8.46) among top directors. Alfred Hitchcock leads in volume (14 movies) but has the lowest average among the top 10.

3. **Votes vs Rating** — Moderate positive correlation (0.495). Movies with very high vote counts rate significantly higher (8.178) than the rest (~7.90).

4. **Decades** — Older decades have fewer but higher-rated movies. Time filters out bad films. 2000s and 2010s dominate in count but have the lowest average ratings.

5. **Missing Revenue** — Missing gross data is heavily concentrated in pre-1980s movies (30-45% missing) due to inconsistent historical record-keeping — not movie quality.

---

## 💾 Save Cleaned Dataset

In [10]:
df.drop(columns=["Has_Gross", "Vote_Range"], inplace=True)
df.to_csv("imdb_top_1000_cleaned.csv", index=False)
print("\nCleaned dataset saved ✅")
print(f"Final shape: {df.shape}")


Cleaned dataset saved ✅
Final shape: (1000, 17)
